controllers + apiserver: enhance context support

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

27a68aee3a4834 introduced context support for events. Creating an event
broadcaster with context makes tests more resilient against leaking goroutines
when that context gets canceled at the end of a test and enables per-test
output via ktesting.
    
#### Special notes for your reviewer:

The context could get passed to the constructor. A cleaner solution is to
enhance context support for the apiserver and then pass the context into the
controller's run method. This ripples up the call stack to all places which
start an apiserver.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
sample controller: enhance context support

27a68aee3a4834 introduced context support for events. Creating an event
broadcaster with context makes tests more resilient against leaking goroutines
when that context gets canceled at the end of a test and enables per-test
output via ktesting.

The New method already had a context, therefore no API changes are needed.

pohly committed:
apiserver + controllers: enhance context support

27a68aee3a4834 introduced context support for events. Creating an event
broadcaster with context makes tests more resilient against leaking goroutines
when that context gets canceled at the end of a test and enables per-test
output via ktesting.

The context could get passed to the constructor. A cleaner solution is to
enhance context support for the apiserver and then pass the context into the
controller's run method. This ripples up the call stack to all places which
start an apiserver.

pohly commented on test/integration/endpoints/endpoints_test.go:
```diff
@@ -162,6 +166,10 @@ func TestEndpointUpdates(t *testing.T) {
 func TestExternalNameToClusterIPTransition(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
A lot of tests calling `kubeapiservertesting.StartTestServerOrDie` need this boiler-plate code although that function internally also does the same thing.

Would it make sense to extend the `kubeapiservertesting.TestServer` struct to include the context, TB and client set interfaces?

Including the client set would allow removing a lot of common code that usually follows, because pretty much every test which creates an API server then also instantiates a client set for it.

The downside is that test code might become a bit weird (?):
```
serverCtx.CoreV1().Pods().Get(serverCtx, ....)
```

pohly replied:
> Would it make sense to extend the kubeapiservertesting.TestServer struct to include the context, TB and client set interfaces?

`test/utils/ktesting` does that, but let's keep using it separate: https://github.com/kubernetes/kubernetes/pull/124093

pohly commented on test/integration/replicationcontroller/replicationcontroller_test.go:
```diff
@@ -111,10 +111,15 @@ func newMatchingPod(podName, namespace string) *v1.Pod {
+	context.Context
+}
+
+func rmSetup(t *testing.T) (tContext, *replication.ReplicationManager, informers.SharedInformerFactory, clientset.Interface) {
```
I've thought a bit more about this and now lean towards putting this into `k8s.io/kubernetes/test/utils/ktesting`.

A test then could start with:
```
func TestSomething(t *testing.T) {
    tCtx := ktesting.Init(t)
...
```

In contrast to the current `ktesting.NewTestContext`, that `Init` would always set up a context with cancellation. Long term, we might also add some configurable per-test deadline.

`tCtx` would be an interface that combines `testing.TB`, `context.Context` and adds `Cancel` (= immediately cancel the context, needed by some tests which start goroutines and wait for them to complete) and `Logger` (= return `logr.Logger) (= convenience function, shorter than `klog.FromContext`).

pohly replied:
If this `ktesting` turns out to be useful, we could move it into a staging repo. Right now, we can only use it internally, but not for unit tests in staging repos.

I would not move this additional API into `klog/v2/ktesting`. That package should remain minimal because of the strict API rules and focus of klog.

pohly replied:
I've pushed [dc0b653](https://github.com/kubernetes/kubernetes/pull/122148/commits/dc0b653657c0b4ea92a515e3027f7869e38c1ba0) which implements what I had in mind.

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -109,7 +111,22 @@ func New(
+	featureGate featuregate.FeatureGate,
+) (*Controller, error) {
+
+	broadcaster := record.NewBroadcaster(record.WithContext(ctx))
```
hmm , but the broadcaster has a shutdown method.

It does not seem right to pass a context to a constructor of an object that have a Run method that already receives the context, that is the context that should shutdown the broadcaster

pohly replied:
It wasn't possible to do it differently with `NewBroadcaster` without breaking that API.

Perhaps here constructing the broadcaster should get delayed until the `Run` method gets invoked? Then the controller `New` methods remain free of the context dependency.

pohly commented on test/integration/events/events_test.go:
```diff
@@ -38,11 +38,11 @@ import (
-	defer cancel()
-	stopCh := ctx.Done()
+	tCtx := ktesting.Init(t)
+	stopCh := tCtx.Done()
```
If others also find `ktesting.TContext` useful (@aojea ?), then I'll rework this PR to use it instead of adding lots of `context.WithCancel` calls.

pohly commented:
/reopen

pohly commented:
/reopen

pohly commented:
/retest

pohly commented:
/retest

pohly commented on staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go:
```diff
@@ -83,13 +84,13 @@ func NewDefaultTestServerOptions() *TestServerInstanceOptions {
 // enough time to remove temporary files.
 func StartTestServer(t Logger, _ *TestServerInstanceOptions, customFlags []string, storageConfig *storagebackend.Config) (result TestServer, err error) {
-	stopCh := make(chan struct{})
+	ctx, cancel := context.WithCancelCause(context.Background())
```
We cannot use ktesting here because it is not a staging repo (yet).

sttts replied:
Add TODO?

pohly replied:
Added.

pohly commented on test/integration/servicecidr/migration_test.go:
```diff
@@ -209,7 +209,7 @@ func TestMigrateServiceCIDR(t *testing.T) {
 	// ServiceCIDR controller
 	tCtx2 := ktesting.Init(t)
-	defer tCtx.Cancel("tearing down ServiceCIDR controller 2")
+	defer tCtx2.Cancel("tearing down ServiceCIDR controller 2")
```
Not related to the reason for this PR, also submitted as a stand-alone fix: https://github.com/kubernetes/kubernetes/pull/124068

aojea replied:
ok, this alreas merged

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -126,6 +127,7 @@ cluster's shared state through which all other components interact.`,
 			return nil
 		},
 	}
+	cmd.SetContext(genericapiserver.SetupSignalContext())
```
🤔  – some go proverb: don't store contexts.

pohly replied:
I know, but sometimes the alternative is breaking the API badly. I bet that was the reason for cobra's decision to add a command context. As that is now part of their API, we should use it.

sttts replied:
I understand why they do it. It's always about late initialization of a context that is wired in constructor code. So you have to cut the link and that only work by accessing a stored context.

pohly replied:
I need to check how Cobra handles sub-commands.

Regarding `genericapiserver.SetupSignalContext` that can be called more than once because it uses `signal.Notify` under the hood and that can be called more than once:
> It is allowed to call Notify multiple times with different channels
> and the same signals: each channel receives copies of incoming signals independently.

pohly replied:
Oh, wait:
https://github.com/kubernetes/apiserver/blob/9b2e61f60b8a7d50ccfb40adad37780cf4038cbf/pkg/server/signal.go#L41

Why? That seems unnecessary.

pohly replied:
> is the context from the root command automatically propagated to the sub commands?

Yes. I checked with this patch:
```patch
diff --git a/cmd/kubectl/kubectl.go b/cmd/kubectl/kubectl.go
index 09c18cfa209..5143366839a 100644
--- a/cmd/kubectl/kubectl.go
+++ b/cmd/kubectl/kubectl.go
@@ -21,12 +21,15 @@ import (
        "k8s.io/kubectl/pkg/cmd"
        "k8s.io/kubectl/pkg/cmd/util"
 
+       genericapiserver "k8s.io/apiserver/pkg/server"
+
        // Import to initialize client auth plugins.
        _ "k8s.io/client-go/plugin/pkg/client/auth"
 )
 
 func main() {
        command := cmd.NewDefaultKubectlCommand()
+       command.SetContext(genericapiserver.SetupSignalContext())
        if err := cli.RunNoErrOutput(command); err != nil {
                // Pretty-print the error and exit with an error.
                util.CheckErr(err)
diff --git a/staging/src/k8s.io/apiserver/pkg/server/signal.go b/staging/src/k8s.io/apiserver/pkg/server/signal.go
index e5334ae4c15..24883d02bfe 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/signal.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/signal.go
@@ -51,7 +51,7 @@ func SetupSignalContext() context.Context {
                os.Exit(1) // second signal. Exit directly.
        }()
 
-       return ctx
+       return context.WithValue(ctx, "foo", "bar")
 }
 
 // RequestShutdown emulates a received event that is considered as shutdown signal (SIGTERM/SIGINT)
```

Then in a sub-command like https://github.com/kubernetes/kubernetes/blob/5fe1e923df1fed37f437241d8d63808121fc3670/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L157 `dlv` confirms that `cmd.ctx` is indeed the context prepared by `SetupSignalContext`:
```
> k8s.io/kubectl/pkg/cmd/run.NewCmdRun.func1() ./staging/src/k8s.io/kubectl/pkg/cmd/run/run.go:157 (hits goroutine(1):1 total:1) (PC: 0x3890707)
...
(dlv) p cmd.ctx
context.Context(*context.valueCtx) *{
	Context: context.Context(*context.cancelCtx) *{
		Context: context.Context(context.backgroundCtx) *(*context.Context)(0xc000522550),
		mu: (*sync.Mutex)(0xc000522560),
		done: (*"sync/atomic.Value")(0xc000522568),
		children: map[context.canceler]struct {} nil,
		err: error nil,
		cause: error nil,},
	key: interface {}(string) "foo",
	val: interface {}(string) "bar",}
```

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -477,12 +477,13 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // |                               |
 // |      HTTPServerStoppedListening (httpServerStoppedListeningCh)
-func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
+func (s preparedGenericAPIServer) Run(ctx context.Context) error {
```
this will probably be used quite a bit in the wild. Worth RunWithContext?

pohly replied:
Okay.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -200,10 +205,12 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
 
 	// start the API server
-	stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+	runCompletedCh := make(chan struct{})
```
I don't like moving the stopCh replacement away. It is crucial in this test. So better create another ctx+cancel here.

pohly replied:
The rationale was that ctx is more like a per-test than per-apiserver context, but I can also leave the setup here.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -423,10 +434,12 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationEnabled(t
 
 	// start the API server
-	stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+	runCompletedCh := make(chan struct{})
```
same as above above not moving it away.

pohly replied:
changed

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -568,10 +584,12 @@ func TestMuxAndDiscoveryComplete(t *testing.T) {
 
 	// start the API server
-	stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+	runCompletedCh := make(chan struct{})
```
same

pohly replied:
Changed.

It's perhaps worth adding that instead of needing three lines of code, ktesting would only be one (`tCtx := ktesting.Init(t)`) because it includes setting up cancellation.

There's no need for a TODO, finding these `ktesting.NewTestContext(t)` calls will be easy.

pohly replied:
https://github.com/kubernetes/kubernetes/pull/122148/files/0a89572aed51477ef9baf4963fc29aa2f1a42d51#r1576754590

sttts commented on staging/src/k8s.io/apiserver/pkg/server/hooks.go:
```diff
@@ -48,8 +49,11 @@ type PreShutdownHookFunc func() error
 	LoopbackClientConfig *restclient.Config
-	// StopCh is the channel that will be closed when the server stops
+	// StopCh is the channel that will be closed when the server stops. Alternatively,
+	// Context can be used. StopCh is provided for present for code that hasn't been updated yet.
```
deprecate?

pohly replied:
No strong opinion. We probably cannot ever be sure that it's not used and thus it has to stay forever, right?

I added a `Deprecated` comment explaining that. We had problems in the past with `Deprecated` comments triggering an avalanche of PRs replacing the deprecated API although there was no real benefit for doing so.

sttts commented:
/assign @aojea 

for the service controller changes.

/assign @tkashem 

for the apiserver startup/shutdown synchronization changes.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -694,7 +697,7 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 // returned if the secure port cannot be listened on.
 // The returned channel is closed when the (asynchronous) termination is finished.
-func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}, shutdownTimeout time.Duration) (<-chan struct{}, <-chan struct{}, error) {
+func (s preparedGenericAPIServer) NonBlockingRun(ctx context.Context, shutdownTimeout time.Duration) (<-chan struct{}, <-chan struct{}, error) {
```
also this one will be used in the wild. Worth a NonBlockingRunContext ?

pohly replied:
Okay.

pohly replied:
Done.

pohly commented:
[Updated](https://github.com/kubernetes/kubernetes/compare/505be9e396a0b8f9413faca0a9cc868767eef880..1db9d01185fe2dfd06c6b0cb092ab0aac71957bd) to address the review feedback.

pohly commented:
/remove-lifecycle rotten

pohly commented:
/retest

pohly commented:
/test pull-kubernetes-integration

cici37 commented:
/triage accepted

aojea commented:
> /assign @aojea
> 
> for the service controller changes.

/lgtm

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -224,6 +218,8 @@ func (c *Controller) Run(ctx context.Context, workers int, controllerManagerMetr
 	defer c.nodeQueue.ShutDown()
 
 	// Start event processing pipeline.
+	c.eventBroadcaster = record.NewBroadcaster(record.WithContext(ctx))
```
does it needs the Shutdown() call if there is a context or is it safe to have both?

pohly replied:
Doing one or the other will cause the broadcaster to stop, so calling Shutdown is optional. Doing both is safe.

pohly replied:
/assign @dims

For approval of this staging/src/k8s.io/cloud-provider/controllers/service/controller.go part.

pohly replied:
BTW, this PR is the first example for the new apidiff job, see https://github.com/kubernetes/kubernetes/pull/122148#issuecomment-2079796389.

pohly commented:
Rebased because of one small conflict in `test/integration/controlplane/transformation/secrets_transformation_test.go`.

tkashem reviewed (commented):
this LGTM, posted a few minor comments/questions, thanks!

pohly commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -200,10 +202,15 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
 	// start the API server
-	stopCh, runCompletedCh := make(chan struct{}), make(chan struct{})
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancelCause(ctx)
```
Will change.

pohly replied:
I also renamed "cancel" -> "stop" for symmetry.

pohly replied:
I need to point out that shadowing the `ctx` variable from `NewTestContext` also has its benefit: it makes it impossible to accidentally use `ctx` instead of `stopCtx`.

pohly commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -200,10 +202,15 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
 		defer close(runCompletedCh)
-		s.PrepareRun().Run(stopCh)
+		if err := s.PrepareRun().RunWithContext(ctx); err != nil {
+			t.Errorf("HTTP server failed: %v", err)
```
Okay

pohly replied:
Done

pohly commented on staging/src/k8s.io/apiserver/pkg/server/hooks.go:
```diff
@@ -151,15 +157,16 @@ func (s *GenericAPIServer) AddPreShutdownHookOrDie(name string, hook PreShutdown
 }
 
-// RunPostStartHooks runs the PostStartHooks for the server
-func (s *GenericAPIServer) RunPostStartHooks(stopCh <-chan struct{}) {
```
I checked with grep.app and only found calls in the apiserver or forks of it.

If you think that we should keep it, then I can keep it as before and add RunPostStartHooksWithContext.`

tkashem replied:
yeah, looks like it could have been a private function of `GenericAPIServer`. I don't think we need `RunPostStartHooksWithContext`. @sttts thoughts?

sttts replied:
Changing the signature should be fine. I don't think we have to be too careful for server functionality (versus client-go and friends).

> yeah, looks like it could have been a private function of GenericAPIServer

Not sure I can follow. This function (whether with context or channel) is part of the public interface of k8s.io/apiserver, and used in projects (I know heavy use in kcp for example).

pohly commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:
```diff
@@ -222,7 +229,7 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
 
 	// signal termination event: initiate a shutdown
-	close(stopCh)
+	cancel(errors.New("shutting down"))
```
See https://github.com/kubernetes/kubernetes/pull/122148/files/0a89572aed51477ef9baf4963fc29aa2f1a42d51#r1576754590

pohly commented:
Comments to address review feedback in https://github.com/kubernetes/kubernetes/compare/0a89572aed51477ef9baf4963fc29aa2f1a42d51..a3c714f482563802bd442cecabe3224073aafb19,
then rebased due to conflict in `secrets_transformation_test.go`.

pohly commented:
/test pull-kubernetes-apidiff

This is a [new job](https://github.com/kubernetes/test-infra/pull/32504) calling [hack/apidiff.sh](https://github.com/kubernetes/kubernetes/pull/122971) - let's see whether it works.

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
`pull-kubernetes-apidiff` works now (kind of - there's one bug in apidiff where it reports false positives for code that isn't changed at all in client-go/utils/workqueue)

The remaining valid changes for this PR are below.

@stts, are those acceptable? See also https://github.com/kubernetes/kubernetes/pull/122148#discussion_r1577787456

@tkashem: did I address your concerns?

 ## ./staging/src/k8s.io/apiextensions-apiserver
Incompatible changes:
- ./pkg/cmd/server.NewServerCommand: changed from func(io.Writer, io.Writer, <-chan struct{}) *github.com/spf13/cobra.Command to func(context.Context, io.Writer, io.Writer) *github.com/spf13/cobra.Command
- ./pkg/cmd/server.Run: changed from func(*k8s.io/apiextensions-apiserver/pkg/cmd/server/options.CustomResourceDefinitionsServerOptions, <-chan struct{}) error to func(context.Context, *k8s.io/apiextensions-apiserver/pkg/cmd/server/options.CustomResourceDefinitionsServerOptions) error
## ./staging/src/k8s.io/apiserver
Incompatible changes:
- ./pkg/server.(*GenericAPIServer).RunPostStartHooks: changed from func(<-chan struct{}) to func(context.Context)
Compatible changes:
- ./pkg/server.PostStartHookContext.Context: added
- ./pkg/server.preparedGenericAPIServer.NonBlockingRunWithContext: added
- ./pkg/server.preparedGenericAPIServer.RunWithContext: added
- context.Context.Deadline: added
- context.Context.Done: added
- context.Context.Err: added
- context.Context.Value: added
## ./staging/src/k8s.io/kube-aggregator
Incompatible changes:
- ./pkg/apiserver.preparedAPIAggregator.Run: changed from func(<-chan struct{}) error to func(context.Context) error
- ./pkg/cmd/server.AggregatorOptions.RunAggregator: changed from func(<-chan struct{}) error to func(context.Context) error
- ./pkg/cmd/server.NewCommandStartAggregator: changed from func(*AggregatorOptions, <-chan struct{}) *github.com/spf13/cobra.Command to func(context.Context, *AggregatorOptions) *github.com/spf13/cobra.Command
## ./staging/src/k8s.io/sample-apiserver
Incompatible changes:
- ./pkg/cmd/server.NewCommandStartWardleServer: changed from func(*WardleServerOptions, <-chan struct{}) *github.com/spf13/cobra.Command to func(context.Context, *WardleServerOptions) *github.com/spf13/cobra.Command
- ./pkg/cmd/server.WardleServerOptions.RunWardleServer: changed from func(<-chan struct{}) error to func(context.Context) error

sttts commented:
Yes, all of those changes are fine.

sttts commented:
/lgtm
/approve

tkashem commented:
/lgtm

pohly commented:
@tkashem: can you re-add LGTM?

I had to rebase because of a minor conflict in `test/integration/framework/test_server.go` (one line, and of course that one got changed on master...).

pohly commented:
There are new test failures after the rebase. Looking...

sttts commented:
/lgtm
